### PR TITLE
Telemetry data initial implementation

### PR DESCRIPTION
**Pull request by astefan**
This is a first iteration of telemetry data support for EQL. Main classes are in, test classes make sense at this point when the hooks will be available in EQL code. Also, same goes for the actual logic that increments the counters.

Part of https://github.com/elastic/elasticsearch/issues/49630.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java [hunk 1]
```diff
@@ -124,6 +124,9 @@ private XPackSettings() {
     public static final Setting<Boolean> FIPS_MODE_ENABLED =
         Setting.boolSetting("xpack.security.fips_mode.enabled", false, Property.NodeScope);
 
+    /** Setting for enabling or disabling eql. Defaults to true. */
+    public static final Setting<Boolean> EQL_ENABLED = Setting.boolSetting("xpack.eql.enabled", true, Setting.Property.NodeScope);
```
Comment (imotov) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java]: This setting already exists in EqlPlugin. Strange that it didn't fail in testing.

File: docs/reference/rest-api/info.asciidoc [hunk 2]
```diff
@@ -119,6 +119,10 @@ Example response:
          "available" : true,
          "enabled" : true
       },
+      "eql" : {
+         "available" : true,
+         "enabled" : true
+      },
```
Comment (imotov) [inline docs/reference/rest-api/info.asciidoc]: That will probably fail in release build, where we disable EQL

**Review by imotov — COMMENTED** [review]
Great addition! Added some comments.

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java (hunk 1 above)
Comment (astefan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java]: @imotov should this setting only exist in EqlPlugin for the moment, and when EQL it's ready to go, move the setting to XPackSettings?

Comment (imotov) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java]: I think the placement of this setting is not as important as having it in one place only and disabling completely in non-snapshot builds. I don't see any issues with moving it in XPackSettings as long as we disable it there for non-snapshot builds.

File: docs/reference/rest-api/info.asciidoc (hunk 2 above)
Comment (imotov) [reply docs/reference/rest-api/info.asciidoc]: That fails in non-snapshot build with the following error:
```
org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT > test {yaml=reference/rest-api/info/line_45} FAILED
    java.lang.AssertionError: Failure at [reference/rest-api/info:16]: $body didn't match expected value:
                             $body: 
                               build: 
                                  date: same [2020-02-04T18:01:06.694307Z]
                                  hash: same [12b24bfa57ae3b282427298cffc735ab9fe5868f]
                            features: 
                             analytics: 
                               available: same [true]
                                 enabled: same [true]
                                   ccr: 
                               available: same [true]
                                 enabled: same [true]
                                enrich: 
                               available: same [true]
                                 enabled: same [true]
                                   eql: 
                               available: same [true]
                                 enabled: expected Boolean [true] but was Boolean [false]
                             flattened: 
                               available: same [true]
                                 enabled: same [true]
                        frozen_indices: 
                               available: same [true]
                                 enabled: same [true]
                                 graph: 
                               available: same [true]
                                 enabled: same [true]
                                   ilm: 
                               available: same [true]
                                 enabled: same [true]
                              logstash: 
                               available: same [true]
                                 enabled: same [true]
                                    ml: 
                               available: same [true]
                                 enabled: same [true]
                            monitoring: 
                               available: same [true]
                                 enabled: same [true]
                                rollup: 
                               available: same [true]
                                 enabled: same [true]
                              security: 
                               available: same [true]
                                 enabled: same [false]
                                   slm: 
                               available: same [true]
                                 enabled: same [true]
                               spatial: 
                               available: same [true]
                                 enabled: same [true]
                                   sql: 
                               available: same [true]
                                 enabled: same [true]
                             transform: 
                               available: same [true]
                                 enabled: same [true]
                               vectors: 
                               available: same [true]
                                 enabled: same [true]
                           voting_only: 
                               available: same [true]
                                 enabled: same [true]
                               watcher: 
                               available: same [true]
                                 enabled: same [true]
                             license: 
                 expiry_date_in_millis: same [1583431616962]
                                  mode: same [trial]
                                status: same [active]
                                  type: same [trial]
                                   uid: same [14927092-6926-43a8-84b9-af6fd185696f]
                             tagline: same [You know, for X]
        at __randomizedtesting.SeedInfo.seed([31BF956B0BB813AA:B9EBAAB1A5447E52]:0)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:405)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:382)
        at jdk.internal.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
        at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
        at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
        at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
        at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
        at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
        at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
        at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Review by imotov — COMMENTED** [review]
It will fail in non-snapshot builds unless we do something with `eql.enabled:true`.

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-windows

**Comment by astefan** [comment]
@elasticmachine update branch

**Review by imotov — APPROVED** [review]
Thanks!

**Pull request merged**